### PR TITLE
Add a Constant helper class.

### PR DESCRIPTION
**Pull request by matanlurey**
Partial progress towards https://github.com/dart-lang/source_gen/issues/169.

`List`, `Map`, and other types are a little trickier, so hoping to get some incremental work first.

### CODE REVIEW

File: lib/src/constants.dart [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+///
+/// Unlike [DartObject.getField], all `readX` methods attempt to access super
+/// classes for the field value if not found.
+abstract class Constant {
+  factory Constant(DartObject object) =>
+      _isNull(object) ? const _NullConstant() : new _Constant(object);
+
+  /// Returns whether this constant represents a `bool` literal.
+  ///
+  /// If `true`, [boolValue] will return either `true` or `false` (not throw).
```
Comment (jakemac53) [inline lib/src/constants.dart]: I don't think you need these lines on the is* getters.

File: lib/src/constants.dart [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+  /// Returns whether this constant represents `null`.
+  bool get isNull;
+
+  /// Reads[ field] from the constant as another constant value.
+  Constant read(String field);
+
+  /// Reads [field] from the constant as a boolean.
+  ///
+  /// If the resulting value is `null`, uses [defaultTo] if defined.
+  bool readBool(String field, {bool defaultTo()});
```
Comment (jakemac53) [inline lib/src/constants.dart]: I am a bit confused why these methods exist in addition to the *Value getters?

File: lib/src/constants.dart [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+      : throw new FormatException('Not a String', _object);
+
+  @override
+  bool get isBool => _object.toBoolValue() != null;
+
+  @override
+  bool get isInt => _object.toIntValue() != null;
+
+  @override
+  bool get isNull => false;
```
Comment (jakemac53) [inline lib/src/constants.dart]: For future proofing it might be best to do `_isNull(_object)` instead?

File: test/constants_test.dart [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+import 'package:build_test/build_test.dart';
+import 'package:source_gen/source_gen.dart';
+import 'package:test/test.dart';
+
+void main() {
+  group('Constant', () {
+    List<Constant> constants;
+
+    setUpAll(() async {
+      constants = (await resolveSource(r'''
```
Comment (jakemac53) [inline test/constants_test.dart]: nit: assign the result of `resolveSource` to a local variable to make this a bit easier to follow. In general I try not to wrap await calls in parens and use local vars instead because it becomes hard to follow.

File: lib/src/constants.dart [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+
+import 'package:analyzer/dart/constant/value.dart';
+
+/// Always returns `null` (used as a default for `defaultTo` methods).
+Null _alwaysNull() => null;
+
+/// Returns whether or not [object] is or represents a `null` value.
+bool _isNull(DartObject object) => object?.isNull != false;
+
+/// Similar to [DartObject.getField], but traverses super classes.
```
Comment (matanlurey) [inline lib/src/constants.dart]: Done.

File: lib/src/constants.dart [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+DartObject _getFieldRecursive(DartObject object, String field) {
+  if (_isNull(object)) {
+    return null;
```
Comment (matanlurey) [inline lib/src/constants.dart]: If you don't mind I'm addressing that in the next PR - the code to check all fields is a little meaty.

Filed https://github.com/dart-lang/source_gen/issues/173.

File: lib/src/constants.dart (hunk 1 above)
Comment (matanlurey) [inline lib/src/constants.dart]: Done.

File: lib/src/constants.dart (hunk 2 above)
Comment (matanlurey) [inline lib/src/constants.dart]: A DartObject can either _be_ a literal or _have_ fields (that are either literals or objects themselves).

Because `Constant` wraps a `DartObject`, it makes sense to make this easier. It also enables (eventually) added support for using a dot notation, or something like:

```dart
constant.readBool('from.another.thing');
```

If you think it's not worth it and we should just have `xValue`, happy to revert though.

File: lib/src/constants.dart (hunk 3 above)
Comment (matanlurey) [inline lib/src/constants.dart]: I think it's impossible to get into a state where that is true, but no harm doing this.

File: test/constants_test.dart (hunk 4 above)
Comment (matanlurey) [inline test/constants_test.dart]: Done.

File: lib/src/constants.dart (hunk 2 above)
Comment (matanlurey) [inline lib/src/constants.dart]: On second thought I agree with you, this makes `defaultTo` not necessary, for example.

**Comment by matanlurey** [comment]
@kevmoo @jakemac53 Let me know if you'd rather this be called `ConstantReader` (in-line with `LibraryReader` in https://github.com/dart-lang/source_gen/pull/172.

**Comment by kevmoo** [comment]
`ConstantReader` – sgtm – @jakemac53 ?
